resources/top_level_overview.markdown

```diff
@@ -0,0 +1,283 @@
+important to note that a combiner basically compresses the data, meaning it
+will eventually have to be decompressed before reducing.
+
+The circular buffer has a configurable threshold, defaulted to 80%. The buffer
+below has `0.8 * 100 MB = 80 MB` of data written, which is the limit in this
+example:
+
+```
+[XXXXXXXX--]
+```
+
+The data is spilled in a separate thread to allow continuation of the map task:
+
+```
+The block in || is being written to disk, x = 10 MB written to memory by mapper
+[|XXXXXXXXX|x-]
+
+Another x = 10 MB written to memory
+[|XXXXXXXXX|xx]
+
+Spill thread completes, clear 80 MB block. 10 MB written to memory by mapper.
+[x-------xx]
+```
+
+Before the write to disk actually occurs, two things happen:
+* The output is split into `R` partitions, again based on the number of
+reduce tasks.
+* The `(key, value)` pairs are sorted, in-memory, using a QuickSort algorithm
+based on (partitionID, key).
+
+In the above case, the rate at which the mapper was writing data to memory was
+less than the rate that it took to spill the block to disk.  However,
+it can be the case where the mapper is writing data faster than the
+spill thread is able to free memory.  In such a scenario, the mapper would be
+blocked until the spill thread has had a chance to clear the buffer.
+
+It's important to note that Hadoop uses zero compressed Long and Int values for
+map output.  Here is a quote from _Hadoop Definitive Guide_, (table 8-2, page 261)
+relating to the tests ran in the _Reverse Engineering Hadoop_ section:
+
+>"Map output materialized bytes" - The number of bytes of map output actually
+written to disk. If map output compression is enabled, this is reflected in the
+counter value.
+
+>"Map output bytes" - The number of bytes of uncompressed output produced by all
+the maps in the job. Incremented every time the collect() method is called on
+the map's OutputCollector.
+
+#### Point 5: Preparing for a reduce task
+MapReduce has the guarantee that the input to every reducer is sorted by key. As
+mentioned in _point 3 & 4_, the system sorts and merges the disjoint spill files
+of individual map tasks and then transfers the map outputs to the reducers.  The
+process where map outputs are fetched by reduce tasks in known as the _shuffle_.
+In other contexts, the shuffle can be taken as the entire process from maps
+producing output, to reducers consuming input.  For our purposes, it helps to
+distinguish the shuffle phase as a discrete step.
+
+An __Event Fetcher__ fetches map task completion events over a network, and then
+adds them to the fetch queue to be consumed by the fetchers themselves.  These
+fetchers are run in separate threads, configurable in
+`mapred.reduce.parallel.copies`, which defaults to five.  The fetchers store the
+map outputs in memory with its own configurable memory buffer.
+
+With Memory-To-Memory enabled, when the amount of map tasks reaches a
```

Comment:
is it really "the amount of map tasks"?  not the amount of map output data?  And "amount" should be "number" then...

Reply:
I think this is actually the number of map task _output files_. I'll look into it more, but this is a parameter disabled by default, and the threshold value is an integer value defaulted to `1000`.  My assumption is to enable this when there are a ton of map tasks being ran as a preemptive measure to the `InMemory` and `OnDisk` mergers.

If this is the case, it should be "number" then, but as I said, I'll dig into this some more.